File: src/_03_BehavioralPattern/_03_03_Interpreter/PlusExpression.js
import PostfixExpression from "./PostfixExpression.js";

export default class PlusExpression extends PostfixExpression {
  /** @type { PostfixExpression } */
  #lhs;
  
  /** @type { PostfixExpression } */
  #rhs;

  /**
   * @param { PostfixExpression } lhs 
   * @param { PostfixExpression } rhs 
   */
  constructor(lhs, rhs) {
    super();
    this.#lhs = lhs;
    this.#rhs = rhs;
  }

  /**
   * @override
   * @param {{ [key: string]: number }} context 
   * @returns { number }
   */
  interpret(context) {
    return this.#lhs.interpret(context) + this.#rhs.interpret(context);
  }
}

File: src/_03_BehavioralPattern/_03_03_Interpreter/java/after/PlusExpression.java
package _03_BehavioralPattern._03_03_Interpreter.java.after;

import java.util.Map;

public class PlusExpression implements PostfixExpression {

  private PostfixExpression lhs;
  private PostfixExpression rhs;

  public PlusExpression(PostfixExpression lhs, PostfixExpression rhs) {
    this.lhs = lhs;
    this.rhs = rhs;
  }
  
  @Override
  public int interpret(Map<Character, Integer> context) {
    return lhs.interpret(context) + rhs.interpret(context);
  }
  
}
